### ISSUE TITLE
[Proposal] Extend the Output Binding block to a general Binding

### ISSUE THREAD

**Issue by RicardoNiepel** [dapr/dapr#1357]
## In what area(s)?

/area runtime
/area docs

> /area operator
> /area placement
> /area test-and-release

## Describe the proposal

Let's extend the Output Binding block to a general Binding building block (or connector as mentioned in #1116).

The current spec defines the Binding contract to include POST/GET/PUT/DELETE
> ```POST/GET/PUT/DELETE http://localhost:<daprPort>/v1.0/bindings/<name>```
> https://github.com/dapr/docs/blob/master/reference/api/bindings_api.md#sending-messages-to-output-bindings

but currently only POST/PUT are implemented ([pkg/http/api.go#L158](https://github.com/dapr/dapr/blob/master/pkg/http/api.go#L158)) and both are used for the same / no distinction between them.

But a lot of the existing bindings could easily support also the others to have a full "connector" feature. This would also fullfil requirements like #1103.

### Interact with Connectors

This endpoint lets you invoke an Dapr connector:
```
POST http://localhost:<daprPort>/v1.0/connectors/<name>
```
> Used with any resource which requires a unique id, like blobs or database entries, it will generate a random id and return it

This endpoint lets you invoke a Dapr connector to replace all current representations of the target resource with the request payload:
```
PUT http://localhost:<daprPort>/v1.0/connectors/<name>/<pk_id>
```

This endpoint lets you invoke a Dapr connector and returns the specified resource:
```
GET http://localhost:<daprPort>/v1.0/connectors/<name>/<pk_id>
```

This endpoint lets you invoke a Dapr connector and deletes the specified resource:
```
DELETE http://localhost:<daprPort>/v1.0/connectors/<name>/<pk_id>
```

#### Generic

| Name | Input<br>Binding | POST Connector | PUT Connector | GET Connector | DELETE Connector | Status |
|------|:----------------:|:-----------------:|:-----:|:----:|:----:|------|
| [HTTP](../../reference/specs/bindings/http.md)           |    | ✅ |✅ |✅ |✅ | Experimental |
| [Kafka](../../reference/specs/bindings/kafka.md)         | ✅ | ✅||| | Experimental |
| [Kubernetes Events](../../reference/specs/bindings/kubernetes.md) | ✅ ||||    | Experimental |
| [MQTT](../../reference/specs/bindings/mqtt.md)           | ✅ | ✅ | |||Experimental |
| [RabbitMQ](../../reference/specs/bindings/rabbitmq.md)   | ✅ | ✅ | |||Experimental |
| [Redis](../../reference/specs/bindings/redis.md)         |     | ✅ |✅ |✅ |✅ | Experimental |
| [Twilio](../../reference/specs/bindings/twilio.md)       |    | ✅ | |||Experimental |

#### Amazon Web Service (AWS)

| Name | Input<br>Binding | POST Connector | PUT Connector | GET Connector | DELETE Connector | Status |
|------|:----------------:|:-----------------:|:-----:|:----:|:----:|------|
| [AWS DynamoDB](../../reference/specs/bindings/dynamodb.md) |    | ✅ |✅ |✅ |✅ | Experimental |
| [AWS S3](../../reference/specs/bindings/s3.md)             |     | ✅ |✅ |✅ |✅ | Experimental |
| [AWS SNS](../../reference/specs/bindings/sns.md)           |    | ✅ |||| Experimental |
| [AWS SQS](../../reference/specs/bindings/sqs.md)           | ✅ | ✅ |||| Experimental |
| [AWS Kinesis](../../reference/specs/bindings/kinesis.md)   | ✅ | ✅ |||| Experimental |

#### Google Cloud Platform (GCP)

| Name | Input<br>Binding | POST Connector | PUT Connector | GET Connector | DELETE Connector | Status |
|------|:----------------:|:-----------------:|:-----:|:----:|:----:|------|
| [GCP Cloud Pub/Sub](../../reference/specs/bindings/gcppubsub.md)  | ✅ | ✅ |||| Experimental |
| [GCP Storage Bucket](../../reference/specs/bindings/gcpbucket.md) |    | ✅ |✅ |✅ |✅ | Experimental |

#### Microsoft Azure

| Name | Input<br>Binding | POST Connector | PUT Connector | GET Connector | DELETE Connector | Status |
|------|:----------------:|:-----------------:|:-----:|:----:|:----:|------|
| [Azure Blob Storage](../../reference/specs/bindings/blobstorage.md)            |     | ✅ |✅ |✅ |✅ | Experimental |
| [Azure EventHubs](../../reference/specs/bindings/eventhubs.md)                 | ✅ | ✅ |||| Experimental |
| [Azure CosmosDB](../../reference/specs/bindings/cosmosdb.md)                   |    | ✅ |✅ |✅ |✅ | Experimental |
| [Azure Service Bus Queues](../../reference/specs/bindings/servicebusqueues.md) | ✅ | ✅ | |||Experimental |
| [Azure SignalR](../../reference/specs/bindings/signalr.md)                     |    | ✅ |||| Experimental |
| [Azure Storage Queues](../../reference/specs/bindings/storagequeues.md)        | ✅ | ✅ |||| Experimental |

**Comment by artursouza** [comment]
This proposed API seems to be a replica of the existing state API. Why the state API cannot be used instead?

**Comment by RicardoNiepel** [comment]
From the docs:
> Application state is anything an application wants to preserve beyond a single session.
> Dapr state stores are key/value stores. To ensure data compatibility, Dapr requires these data stores follow a fixed key scheme.
> For general states, the key format is: ```<Dapr id>||<state key>```
> For Actor states, the key format is: ```<Dapr id>||<Actor type>||<Actor id>||<state key>```

Essentially the Dapr State Store building block enables stateful services and stateful actors.

The **proposal** above defines a general way of connecting Dapr applications with external systems.
The Pub/Sub and State Store building blocks are just specific manifestations of it, enabling the two scenarios with additional functionality, but also bring with them special requirements for the external systems underneath. 

In my eyes that's exactly the starting point of discussion that has led to #1116:
The proposed transition from a narrow **Output Binding** model to a generic **Connector** model.

## Looking around in the industry/community: <br>Azure Functions as example
Azure Functions distinguishes between Triggers, Inputs and Outputs.
Fuctions Triggers are very similiar to Dapr Input Binding and Functions Outputs are similiar to Dapr Output Binding.

For Functions Inputs Dapr does not have any equivalent right now - this would be the GET Connectors from the proposal.

Looking at Functions Inputs/Outputs, some of them are already supporting additional Scenarios like DELETE, Query etc.:
- bind ```DocumentClient``` at [CosmosDB Input](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-cosmosdb-v2-input?tabs=csharp#http-trigger-get-multiple-docs-using-documentclient) enables raw SDK access but still leverage the client creation inluding secret handling of Azure Functions
- bin ```CloudBlockBlob``` at [Blob Storage Output](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-storage-blob-output?tabs=csharp#usage) enables raw SDK access but still leverage the client creation inluding secret handling of Azure Functions

Looking at the community / most asked questions, a lot of developers are asking how they can use Azure Functions Triggers, Inputs and Outputs for all their needs without the need to use the SDKs directly.

### Possible conclusion for Dapr
- Having a GET Connector in addition to the Dapr Triggers supports scenarios where multiple inputs are used inside the application processing, but only a port of these are triggering the executions
- Having PUT and DELETE Connectors enables developers to leverage Dapr in more scenarios, benefit from the abstractions and Dapr features and minimizes their need to use the SDKs directly

## Cannot the State Store building block used in these situations?
Dapr stores any state tightly bounded to the application (```<Dapr id>||<state key>```) or the actor instance (```<Dapr id>||<Actor type>||<Actor id>||<state key>```).

In scenarios where a connection to an external system - _outside of the application scope or whole system scope_ - is required, the State Store building block cannot be used, it enables only stateful services and stateful actors.

### Example 1: Generic HTTP Connector
Enables using Dapr for connecting to any external API.

- `url` is the HTTP endpoint to invoke
- `allowedMethods` defines which HTTP verbs are supported at this HTTP connector
- `authorizationToken` (optional) defines the authorization bearer token used
- `retryInterval` (optional) in milliseconds
- `retryPattern` (optional) can be either *linear* or *exponential*

```yml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: <name>
spec:
  type: connector.http
  metadata:
  - name: url
    value: http://external-api.com
  - name: allowedMethods
    value: "POST,PUT,GET,DELETE"
  - name: authorizationToken
    secretKeyRef:
      name: ext-api-secret
      key: ext-api-token
  - name: retryInterval
    value: 50
  - name: retryPattern
    value: exponential
auth:
    secretStore: azurekeyvault
```

This is clearly no State Store use case, but a valid Connector one.

**Dapr can provide**
- Add abstraction
- Handles retries and failure recovery
- Add Observability for outgoing calls
- Frees applications of handling URLs, secrets etc.
- Switch between bindings at runtime time
- Enable portable applications where environment-specific bindings are set-up and no code changes are required

### Example 2: Generic Azure Blob Storage Connector <br>(AWS S3 / GCP Storage Bucket as well)
Enables the usage of Dapr for all blob interactions.

- `storageAccount` is the Blob Storage account name
- `storageAccessKey` is the Blob Storage access key
- `container` is the name of the Blob Storage container to write to
- `allowedMethods` defines which HTTP verbs are supported 
```yml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: <name>
spec:
  type: connector.azure.blobstorage
  metadata:
  - name: storageAccount
    value: myStorageAccountName
  - name: storageAccessKey
    secretKeyRef:
      name: storage-secret
      key: storage-prim-key
  - name: container
    value: container1
  - name: allowedMethods
    value: "PUT,GET,DELETE"
auth:
    secretStore: azurekeyvault
```

This is no State Store use case if the state stored goes behind the scope of the application.
E.g. this could be
- accessing external blob storage, only have read access
- storing blobs for external consumption

**Dapr can still provide all benefits as above**

### Example 3: Databases...
The examples go on in the database space - very similiar to above, where this is no State Store use case:
- inserting/manipulating data inside a reporting database, which is consumed by someone else / a reporting system


I look forward to your feedback
/cc @AaronCrawfis @amanbha @yaron2 @msfussell

**Comment by AaronCrawfis** [comment]
I agree this is the direction output bindings should take: extend the current POST/PUT model to include GET/DELETE.

Just as you outlined it could be an easy way for developers to get the retry/failure logic, observability, and secret management (plus more) out of the box for free. Just plug in a connector (or binding, naming is still up in the air) and you're up and running with an external service.

**Comment by artursouza** [comment]
First, my point is not that the state store should work as-is but by extending it. The retry logic can be implemented in the state store APIs too. Same with secret reference. The only thing that is different from the state store in this requirement is the composite key - in this case, the state key can the an auto-generated message id.

The examples of http connector and blob store can be achieved via the state store API as well. The block storage case above seems to be a perfect fit for the state store API, actually.

My point is that the binding API will become almost (if not exactly) a clone of the state API - just with a different name. Maybe we should merge binding and state store into a single API.

**Comment by RicardoNiepel** [comment]
> "Maybe we should merge binding and state store into a single API."

Exactly - let us also think about the current overhead we have to implement some external connectors in all three building blocks just with little differences:
1. Binding
2. State Store
3. PubSub

This is also in line with my statement above
> "[This] defines a general way of connecting Dapr applications with external systems.
The Pub/Sub and State Store building blocks are just specific manifestations of it,"

Migrating from three different building blocks to only one main binding ("conectors") block, where  StateStore and PubSub can leverage the underlying binding code, would be a big change, but minimizes a lot of work going forward of supporting more external systems.

**Comment by yaron2** [comment]
I agree with making the bindings API into connectors with more HTTP verb operations to increase the number of supported operations, but I do not see it becoming a state store API (or Pub Sub) API for several reasons.

I will be talking about Bindings and State below, but the same is true for Bindings and every other building block.

Dapr follows _explicit_ API design, being somewhat of a black box (by the virtue of running out of process and employing non-typed Restful APIs) it should carefully follow the **principle of least surprise** as far as the user is concerned.

So far in the discussion only the RESTful API model was taken into account, but there is also a service model in the form of a gRPC proto service.

Luckily, proto services help us see the whole picture as they are typed. Consider the following API:

```
 rpc InvokeBinding(BindingRequest) returns (BindingsResponse) {}
```
Now consider the ```BindingRequest``` object and what it would contain to satisfy both bindings (an external system with loosely coupled data) and a state store, an explicit API model with very specific features and properties like `consistency`, `concurrency`, and more showing up in the future.

Clearly, it is not possible to converge multiple, very different API models into a single building block without either creating a generic way of supplying metadata which breaks the element of least surprise, or putting every possible field into the same envelope, which creates clutter and becomes unclear to the user about how different fields play out together.

Going back to a RESTFul example, convergence wouldn't work here either, even with JSON being a schema-less object model, as the problem is with the API model itself.

Unifying bindings and state means that the API surface and ability to extend, enrich and change the State API is _**severely**_ limited, and will result in thinking about how to fit specific API models into a single one.

Take the case of a transactions API, for example.
With using the same API for both building blocks, the following API becomes impossible:

```
v1.0/state/<state-store-name>/transaction/<key>
```

Unfortunately, the problem of breaking the element of least surprise and non-explicitness becomes visible with a RESTful model too.

Carrying the features (such as transaction for example) as part of the payload causes confusion to the developer who looks at the `connector/binding` url and cannot figure out what is the RESTFul operation being attempted here without inspecting the payload, which essentially becomes a bag of properties. (And is also not possible on `GET` requests).

One would then think about the query parameters to denote building block specific operations, but query parameters are very much non-ideal means of expressing an API.

The biggest concern with using a single API model for multiple building blocks is the fear of lock in, the very same feat Dapr offers developers.

As Dapr grows, its building blocks grow, not only in number but in features. by trying to enforce a single API model for multiple building blocks, we inevitably find ourselves thinking about how to make the State API work with the Bindings API so that they can share a RESTFul API model, instead of focusing on the business domain problem at hand and how to best express it in an API model that is natural to the problem at hand and is explicit.

That being said, in terms of the new suggested design for bindings, I think it is a great design to make bindings/connectors more extensible and offers a very powerful way to contacting _external_ systems.

**Comment by amanbha** [comment]
I agree with not merging building blocks and enhancing bindings/connectors block.

**Comment by artursouza** [comment]
Thanks for the thoughtful feedback on this.

I agree that merging is not the right approach mainly because it will lock those together. In the mean time, the bindings API will be very similar to the state API and the specific scenario in blob store could be addressed in state store API.

I am outnumbered here :), so I will disagree and commit.

— Artur Souza
________________________________

**Comment by yaron2** [comment]
Closed with #1116

**Comment by artursouza** [comment]
Blob storage should be the first scenario (component) to have new actions implemented.
